app/api/attest/twitter/route.ts

```diff
@@ -34,6 +46,19 @@ export async function POST(req: NextRequest) {
   if (await attestedTwitter(walletAddress)) {
     return NextResponse.json({ error: 'User has attestation' }, { status: 400 })
   }
-  const signedResponse = await signTwitter(session.user?.linkedAccounts?.twitter.split(':')[0], walletAddress);
-  return NextResponse.json({ signedResponse })
+
+  const lock = new Lock({
+    id: key,
+    lease: 5000,
+    redis: redis,
+  })
+
+  if (await lock.acquire()) {
```

Comment:
You might  need to acquire this lock before reading from the cache here:

```typescript
const current = await redis.get(key);
```

Or it is possible to have the following race condition with 2 requests being processed in parallel:


- request 1 checks for the cache and sees that the attestation is not there, it then proceeds to check the attestation in the blockchain (`if (await attestedTwitter(walletAddress))`).
- since the first request is paused waiting for the blockchain check, request 2 will also proceed to checking the attestation in the blockchain
- both requests will acquire the lock reach this block


My suggestion is to acquire the lock right before the first cache check, and surround everything else in a `try/finally` block, with the finally block release the lock. This avoids having to release the lock in every early return.

Reply:
Thanks for this, I went with your suggestion.